dag_in_context/src/interval_analysis.egg

```diff
@@ -249,4 +265,73 @@
        (= (IntB hi) (hi-bound v))
       )
       ((set (hi-bound lhs) (IntB  hi)))
-      :ruleset interval-analysis)
\ No newline at end of file
+      :ruleset interval-analysis)
+
+
+
+; If the If takes a tuple
+(rule (
+        ; expr < value
+        (= pred (Bop (LessThan) expr value))
+        (= if_e (If pred inputs then else))
+        ; the left operand of the < is an input to the if region
+        (= expr (Get inputs i))
+        ; the right operand of the < has an upper bound
+        (= (IntB v) (hi-bound value))
+        ; context node inside the if region
+        (= ctx (InContext (InIf true pred inputs) (Arg ty)))
+        (HasType inputs ty)
+        ; usage of the value x inside the if region
+        (Get ctx i)
+      )
+      ((set (hi-bound (Get ctx i)) (IntB (- v 1)))) ; expr < value was true, so we know expr < (hi-bound value) - 1
+      :ruleset interval-analysis)
+(rule (
+        ; expr < value
+        (= pred (Bop (LessThan) expr value))
+        (= if_e (If pred inputs then else))
+        ; the left operand of the < is an input to the if region
+        (= expr (Get inputs i))
+        ; the right operand of the < has a lower bound
+        (= (IntB v) (lo-bound value))
+        ; context node inside the if region
+        (= ctx (InContext (InIf false pred inputs) (Arg ty)))
+        (HasType inputs ty)
+        ; usage of the value x inside the if region
+        (Get ctx i)
+      )
+      ((set (lo-bound (Get ctx i)) (IntB v))) ; x < y was false, so we know expr >= (lo-bound value)
+      :ruleset interval-analysis)
+
+(rule (
+        ; value < expr
+        (= pred (Bop (LessThan) value expr))
+        (= if_e (If pred inputs then else))
+        ; the right operand of the < is an input to the if region
+        (= expr (Get inputs i))
+        ; the left operand of the < has a lower bound
+        (= (IntB v) (lo-bound value))
+        ; context node inside the if region
+        (= ctx (InContext (InIf true pred inputs) (Arg ty)))
+        (HasType inputs ty)
+        ; usage of the value x inside the if region
+        (Get ctx i)
+      )
+      ((set (lo-bound (Get ctx i)) (IntB (+ v 1)))) ; value < expr was true, so we know expr >= (lo-bound value) + 1
```

Comment:
They're equivalent, so why not write expr>lo value instead of expr>=lo value + 1?

Reply:
I recall you were working on these sort of normalization rules- did they end up having a problem?

Reply:
I wrote normalization rules that phrase everything in terms of < (not <=, >, >=)

Reply:
I get it now, it's because the lo-bound is inclusive. Just change the comment in the previous convo.